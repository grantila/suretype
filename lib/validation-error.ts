import type * as Ajv from "ajv"

import { getPrettify } from "./ajv-errors.js"
import { SuretypeOptions, getSuretypeOptions } from "./options.js"


export type ErrorHook = ( err: ValidationError ) => void;

let errorHook: ErrorHook | undefined = undefined;

export function setErrorHook( fn?: ErrorHook | undefined )
{
	errorHook = fn;
}

export interface ValidationErrorData
{
	errors: Array< Ajv.ErrorObject >;
	explanation?: string;
}
export interface ValidationResultInvalid extends ValidationErrorData
{
	ok: false;
}
export interface ValidationResultValid
{
	ok: true;
	errors?: Array< Ajv.ErrorObject >;
	explanation?: string;
}
export type ValidationResult = ValidationResultInvalid | ValidationResultValid;

export interface ExplanationOptions extends SuretypeOptions
{
	schema: unknown;
	data: unknown;
	noFallback?: boolean;
}

export class ValidationError extends Error implements ValidationErrorData
{
	// @ts-ignore
	explanation: string;

	constructor(
		public errors: Array< Ajv.ErrorObject >,
		options: ExplanationOptions
	)
	{
		super( 'Validation failed' );

		makeExplanationGetter( this, 'explanation', errors, options );

		errorHook?.( this );
	}
}

function makeExplanation(
	errors: Array< Ajv.ErrorObject >,
	{
		schema,
		data,
		colors = getSuretypeOptions( ).colors,
		location = getSuretypeOptions( ).location,
		bigNumbers = getSuretypeOptions( ).bigNumbers,
		noFallback = false
	}: ExplanationOptions
)
{
	if ( schema && typeof schema === 'object' )
		return getPrettify( )( {
			errors,
			schema,
			data,
			colors,
			location,
			bigNumbers,
		} );
	else if ( !noFallback )
		return JSON.stringify( errors, null, 2 );
	else
		return undefined;
}

export function makeExplanationGetter< T extends { }, P extends string >(
	target: T,
	property: P,
	errors: Array< Ajv.ErrorObject >,
	options: ExplanationOptions
)
: T & { [ p in P ]: string; }
{
	let cache: string | undefined = undefined;
	Object.defineProperty( target, property, {
		get( )
		{
			if ( cache !== undefined )
				return cache;

			cache = makeExplanation( errors, options );
			return cache;
		}
	} );

	return target as any;
}
