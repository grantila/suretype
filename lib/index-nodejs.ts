import "./json-schema-nodejs.js"
import "./ajv-errors-nodejs.js"
export * from "./index-core.js"


// Patch the ValidationError to make Node.js console printing prettier

import { inspect } from "node:util"
import { setErrorHook, ValidationError } from "./validation-error.js"

( ValidationError.prototype as any )[ inspect.custom ] = function( )
{
	return ( this as ValidationError ).explanation;
}

const debouncedExplanations = new WeakSet< ValidationError >( );

setErrorHook( ( err ) =>
{
	const { message, stack } = err;
	Object.defineProperties( err, {
		message: {
			get( this: ValidationError )
			{
				if ( debouncedExplanations.has( this ) )
					return message;
				debouncedExplanations.add( this );
				setImmediate( ( ) => debouncedExplanations.delete( this ) );
				return message + "\n" + this.explanation;
			}
		},
		stack: {
			get( this: ValidationError )
			{
				if ( debouncedExplanations.has( this ) )
					return stack;
				debouncedExplanations.add( this );
				setImmediate( ( ) => debouncedExplanations.delete( this ) );
				return this.explanation + "\n" + stack;
			}
		},
	} );
} );
