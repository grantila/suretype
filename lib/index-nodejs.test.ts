import { inspect } from "node:util"

import { v, compile, validate, isValid, ensure } from "./index-nodejs.js"
import { ValidationError } from "./validation-error.js"


const getError = ( fn: ( ) => any ): Error =>
	{
		try
		{
			fn( );
		}
		catch ( err: any )
		{
			return err;
		}
		throw new EvalError( 'Function should have thrown' );
	};

describe( "nodejs", ( ) =>
{
	it( "should inject decorated console output for ValidationError", ( ) =>
	{
		expect( ( ValidationError.prototype as any )[ inspect.custom ] )
			.toStrictEqual( expect.any( Function ) );
	} );

	it( "should get decorated ValidationError", ( ) =>
	{
		const schema = v.object( { foo: v.number( ) } );
		const validator = compile( schema, { ensure: true } );
		const error = getError( ( ) => validator( { foo: "42" } ) );

		expect( inspect( error ) ).toMatchSnapshot( );
	} );
} );

describe( "short-hands", ( ) =>
{
	const schema = v.object( {
		firstName: v.string( ).required( ),
		lastName: v.string( ).required( ),
	} );

	const valueValid = { firstName: 'Foo', lastName: 'Bar' } as unknown;
	const valueInvalid = { firstName: 'Foo', lastName: 42 } as unknown;

	it ( "validate", ( ) =>
	{
		const validationValid = validate( schema, valueValid );
		const validationInvalid = validate( schema, valueInvalid );

		expect( validationValid.ok ).toBe( true );
		expect( validationInvalid.ok ).toBe( false );
		expect( validationInvalid.errors?.length ).toBe( 1 );
		expect( validationInvalid.explanation ).toContain( "lastName" );
	} );

	it ( "isValid", ( ) =>
	{
		const True = isValid( schema, valueValid );
		const False = isValid( schema, valueInvalid );

		if ( True )
		{
			// This compiles because of type-guarded return
			expect( valueValid.firstName ).toBe( 'Foo' );
		}

		expect( True ).toBe( true );
		expect( False ).toBe( false );
	} );

	it ( "ensure", ( ) =>
	{
		const user = ensure( schema, valueValid );

		// This compiles because of type-guard
		expect( user.firstName ).toBe( 'Foo' );

		expect( ( ) => ensure( schema, valueInvalid ) )
			.toThrowErrorMatchingSnapshot( );
	} );
} );
