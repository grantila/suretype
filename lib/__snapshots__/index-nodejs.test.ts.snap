// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nodejs should get decorated ValidationError 1`] = `
"The type of the .foo value should be number, e.g. 42 (as number)

  1 | {
> 2 |   "foo": "42"
    |          ^^^^ Replace this with e.g. 42 (as number)
  3 | }"
`;

exports[`short-hands ensure 1`] = `
"Validation failed
The type of the .lastName value should be string, e.g. "42" (as string)

  1 | {
  2 |   "firstName": "Foo",
> 3 |   "lastName": 42
    |               ^^ Replace this with e.g. "42" (as string)
  4 | }"
`;
